Sources/GRPC/ConnectionManager.swift

```diff
@@ -282,51 +281,86 @@ internal class ConnectionManager {
     self.logger = logger
   }
 
-  /// Returns a future for a connected channel.
-  internal func getChannel() -> EventLoopFuture<Channel> {
-    return self.eventLoop.flatSubmit {
-      let channel: EventLoopFuture<Channel>
-
-      switch self.state {
-      case .idle:
-        self.startConnecting()
-        // We started connecting so we must transition to the `connecting` state.
-        guard case let .connecting(connecting) = self.state else {
-          self.invalidState()
-        }
-        channel = connecting.readyChannelPromise.futureResult
+  /// Get the multiplexer from the underlying channel handling gRPC calls.
+  /// if the `ConnectionManager` was configured to be `fastFailure` this will have
+  /// one chance to connect - if not reconnections are managed here.
+  internal func getHTTP2Multiplexer() -> EventLoopFuture<HTTP2StreamMultiplexer> {
+    switch self.configuration.callStartBehavior.wrapped {
+    case .waitsForConnectivity:
+      return self.getHTTP2MultiplexerPatient()
+    case .fastFailure:
+      return self.getHTTP2MultiplexerOptimistic()
+    }
+  }
 
-      case let .connecting(state):
-        channel = state.readyChannelPromise.futureResult
+  /// Returns a future for the multiplexer which succeeded when the channel is connected.
+  /// Reconnects are handled if necessary.
+  private func getHTTP2MultiplexerPatient() -> EventLoopFuture<HTTP2StreamMultiplexer> {
+    let multiplexer: EventLoopFuture<HTTP2StreamMultiplexer>
 
-      case let .active(state):
-        channel = state.readyChannelPromise.futureResult
+    switch self.state {
+    case .idle:
+      self.startConnecting()
```

Comment:
Would it be easier for `startConnecting` to just return the future you want?

Reply:
Probably not as in `getHTTP2MultiplexerOptimistic` the one you want is different.  I guess the logic could be pushed in but then the decision point wouldn't match for the other cases.